pyEpiabm/pyEpiabm/sweep/travel_sweep.py

```diff
@@ -0,0 +1,261 @@
+                    # Add new microcell
+                    microcells_to_choose_dict[possible_microcell] = \
+                        len(possible_microcell.persons)
+
+        # Assign to microcell and household in existing population
+        for person in self.initial_cell.persons:
+            selected_microcell = random.choice(list(
+                microcells_to_choose_dict.keys()))
+            # Assign person to microcell and microcell to person
+            selected_microcell.add_person(person)
+            person.microcell = selected_microcell
+            r = random.random()
+            if r < self.travel_params['prob_existing_household']:
+                # Assign to existing household
+                selected_household = random.choice(
+                    selected_microcell.households)
+                selected_household.add_person(person)
+            else:
+                # Create new household
+                selected_microcell.add_household([person])
+
+    def check_leaving_individuals(self, time, person):
+        """
+        Check if individuals travel_end_time is reached. If interventions
+        are active and travelling individual is in isolation and/or quarantine
+        individuals leave after their travel_end_time is reached and they are
+        out of isolation and/or quarantine.
+
+        Parameters
+        ----------
+        time : float
+            Simulation time
+        Person : Person
+            Instance of Person class
+
+        """
+        if (hasattr(person, 'travel_end_time')) and \
+                (time > person.travel_end_time):
+            if all([hasattr(person, 'isolation_start_time'),
+                    hasattr(person, 'quarantine_start_time')]):
+                if (person.isolation_start_time is None and
+                        person.quarantine_start_time is None) or \
+                        (person.isolation_start_time <= time and
+                         person.quarantine_start_time <= time):
+                    return True
+            elif hasattr(person, 'isolation_start_time'):
+                if (person.isolation_start_time is None) or \
+                        (person.isolation_start_time <= time):
+                    return True
+            elif hasattr(person, 'quarantine_start_time'):
+                if (person.quarantine_start_time is None) or \
+                        (person.quarantine_start_time <= time):
+                    return True
+            else:
+                return True
+        return False
+
+    def remove_leaving_individuals(self, time):
+        """
+        Remove individuals after their travel_end_time is reached
+        from cell, microcell and household.
+
+        Parameters
+        ----------
+        time : float
+            Simulation time
+
+        """
+        someone_is_leaving = False
+        for person in self.travellers:
+            if self.check_leaving_individuals(time, person):
```

Comment:
Why not just remove the person from the travellers list at this point, instead of regenerating the entire list?

Reply:
If you remove them in the loop you will never remove the last person as the list is getting shorter.

Reply:
Hmm yes I see the issue - I do think this construction is very unwieldy though! There is actually extended discussion about the issue on [StackExchange](https://stackoverflow.com/questions/1207406/how-to-remove-items-from-a-list-while-iterating#:~:text=If%20you%20want%20to%20delete,end%20index%20after%20each%20deletion.).

I think my preferred solutions are:
1.) Make a copy of the list (you can either iterate over the copy and change the original, or iterate over the original and make the copy by appending to a new list as you go*.
2.) Iterating through list backwards (with `[::-1]` or `reversed()`)

Either of these save you making twice as many (potentially costly) calls to `check_leaving_individuals()` as you need.

*The former is more space efficient, while the latter is more time efficient - I think the latter is also less likely to misbehave so probably have a slight preference for that, but happy for you to go with either

Reply:
Thanks! I changed it to using `reversed()`.

Reply:
Great work!